### PR TITLE
Remove GSL dependency and add dependency on Eigen

### PR DESCRIPTION
**Pull request by lornat75**
Removes GSL dependency from libYARP_math and libYARP_sig. Interoperability with GSL is delegated to libYARP_gsl. libYARP_math can now be compiled with GSL (for legacy reasons) or Eigen.

### CODE REVIEW

File: src/libYARP_math/src/math.cpp [hunk 1]
```diff
@@ -7,34 +7,12 @@
...
-#include <gsl/gsl_vector.h>
-#include <gsl/gsl_vector_double.h>
-#include <gsl/gsl_blas.h>
 
-#include <gsl/gsl_linalg.h>
-
-#include <gsl/gsl_version.h>
-#include <gsl/gsl_eigen.h>
+#include <cmath>
+#include <gsl/gsl_math.h>
```
Comment (traversaro) [inline src/libYARP_math/src/math.cpp]: @lornat75 what is the rationale of this include? `math.cpp` is supposed to be GSL independent now, and any M_PI problem on Windows can be solved independently  from GSL, see https://github.com/robotology/QA/issues/81 .

**Comment by traversaro** [comment]
See https://github.com/robotology/yarp/issues/840 .

**Comment by drdanz** [comment]
@traversaro added [WIP] tag, please remove it when it is ready for merging

File: src/libYARP_math/src/math.cpp (hunk 1 above)
Comment (traversaro) [inline src/libYARP_math/src/math.cpp]: Apparently this is causing failure on system without gsl : https://ci.appveyor.com/project/robotology/yarp/build/appveyor.1172/job/kvwmlqcf55bh5hbw#L653 . Working on solving the problem in it.

**Comment by traversaro** [comment]
`libYARP_math` with Eigen compilation and testing have been successfully added to AppVeyor, that is now failing because apparently ` dev::FrameTransformClientTest` does not work on Windows, see https://ci.appveyor.com/project/robotology/yarp/build/appveyor.1178/job/1na8hqf27syhblv7 .

**Comment by drdanz** [comment]
Rebased over latest `devel`
@traversaro Since now icub-main has a `devel` branch as well, I think we are ready to merge this, right?

**Comment by traversaro** [comment]
I have to create the matching `icub-main` PR.

**Comment by traversaro** [comment]
Depends on https://github.com/robotology/yarp/pull/919 being merged.

**Comment by traversaro** [comment]
@drdanz  I updated the version and the documentation. 

The only two things missing are : 
 * https://github.com/robotology/homebrew-cask/issues/4
 * https://github.com/robotology/yarp/issues/646

Given that Eigen is currently not included in the installer, Windows users need to manually install Eigen, but I provided detailed instruction on how to do it in  https://github.com/robotology/yarp/pull/896/files#diff-cc43d8d901c5d227b116afb19210e188R27 .

**Comment by drdanz** [comment]
We could also consider adding it to external...

**Comment by traversaro** [comment]
@drdanz I would strongly prefer to avoid that, unless it is a (really) temporary situation, or it is an alternative solution that does not delays the introduction of Eigen in YARP binary installer before YARP 2.3.68 . 

In CoDyCo we have developed several software (such as iDynTree) whose only dependency not included in the YARP/ICUB binaries installer is Eigen. Including Eigen as an external in YARP would not simplify the process of compiling this software in Windows (see https://github.com/robotology/yarp/issues/646). 

@lornat75 What do you think about including Eigen in the dependencies installer?

**Comment by pattacini** [comment]
I definitely vote for incorporating `Eigen` in Windows installer, otherwise we'll really introduce a breach in the system.

**Comment by drdanz** [comment]
+1 For eigen in the installer, but I was proposing to add it to external, since the installer will not be available until we release yarp 2.3.68

**Comment by pattacini** [comment]
@drdanz I think this would represent a very good reason to ship a new release as soon as we merge `Eigen` into `yarp` :smile:

**Comment by drdanz** [comment]
I strongly recommend a small "freeze" period before the release, in order to have some time to test the latest new features.
Anyway, @traversaro #919 is merged, do you think we can merge this now?

**Comment by pattacini** [comment]
Yes, @drdanz you're right in the sense we're using `devel` now.

What about in the future using feature branches to test code too, and only after that, merging in `devel`?

**Comment by drdanz** [comment]
> What about in the future using feature branches to test code too, and only after that, merging in devel?

I think that's the reason for the `devel` branch... You put stuff that needs testing in `devel`, and people will start using it, and then it will end in `master` whenever it is "tested enough®", the problem is that is we never stop adding features to `devel`, the latest features will always be too new to declare them "stable"...

An alternative would be to create a `release` branch from `devel`, freeze it and deploy bug fixes to the `release` branch, and new features to `devel`. When we are ready, the `release` branch is merged to master instead of the `devel` branch and we tag the release... What do you think?

**Comment by drdanz** [comment]
Also IIRC, we decided to remove GLS completely for the next release, and the flag `CREATE_LIB_MATH_USING_GSL` is supposed to be temporary, for a transition period on devel, am I remembering wrong?

**Comment by francesco-romano** [comment]
@drdanz that is a workflow I like for "almost-production" code.  (http://nvie.com/posts/a-successful-git-branching-model/)
The question is (at least, this is what @traversaro always ask me): are we capable of supporting this workflow?

**Comment by pattacini** [comment]
Let me better explain my concern.

At the moment it seems we've been using `features branches` only for developing code without testing it at run-time (we usually check only if it compiles). Is that right? Correct me if I'm wrong, I might be missing crucial points here.

Testing the new feature when is in `devel` does mean that we have to verify the whole ensemble - which is good - but if something bad happens we are already at the `devel` stage, that is the code running on our robots, and doing rollback might be not always easy when we're rushing and the system is messed up.

On the contrary, if we **rebase the feature** on top of `devel` and carry out a first round of tests on a single dedicated platform I think we'll be safer since we haven't merged into `devel` straightaway.

Of course, this means that there must be an **assignee** to take care of this preliminary tests (who might be most likely the author of the new feature) and that since this procedure is somewhat **time consuming** we could reserve it just to big **changes** :wink: 

Once merged into `devel`, well the whole code will be under test as usual. In summary, I'd propose to **introduce preliminary tests just for big changes**.

Now, what you're saying here
> never stop adding features to devel, the latest features will always be too new to declare them "stable"

is also definitely true! However, in my opinion it would be better to increase the number of releases rather than getting the workflow more complicated with the `release` branch.

**Comment by claudiofantacci** [comment]
My 2 cents.
I agree with @pattacini to

> rebase the feature on top of devel and carry out a first round of tests on a single dedicated platform I think we'll be safer since we haven't merged into devel straightaway. 

implying that

> there must be an assignee to take care of this preliminary tests (who might be most likely the author of the new feature)

which is also compliant with, and by my understanding the reason of adopting, the [workflow](http://nvie.com/posts/a-successful-git-branching-model/) linked by @francesco-romano.

**Comment by drdanz** [comment]
> At the moment it seems we've been using features branches only for developing code without testing it at run-time (we usually check only if it compiles). Is that right?

We test that YARP compiles and that *some parts* (the ones tested by unit tests) work and do not contain memory leaks on several platforms, compilers and combination of CMake flags... That's in my opinion already a lot of testing for feature branches... Also branches that require testing on the robots are already usually tested before merging.
The problem in my opinion is the testing of the dependencies. While I can build icub-main on my pc, building icub main and all the other dependencies, superbuilds, etc would require an unreasonable amount of time, so either we find a way to do all these tests automatically, or we push them on devel and wait for feedback of other users (that usually arrives in about 5seconds for MSVC thanks to @pattacini :sob:).

For example here we are modifying YARP to use Eigen instead of GSL. Everything seems to be working, but, for example, I don't think that Eigen is tested by Appveyor, therefore I don't know what will happen on MSVC. Moreover there are several dependencies that use YARP_math, we can't possibly test all of them before merging this pull request, we must accept that after merging this there might be some related issue... `devel` is by definition an *unstable* branch, and therefore this kind of issue is IMO acceptable, `master` on the other hand is *stable* and this kind of issue is unacceptable.

> Of course, this means that there must be an assignee taking care of these preliminary tests (who might be most likely the author of the new feature) and that since this procedure is somewhat time consuming we could reserve it just to big changes :wink:

That's absolutely not doable, first of all because people who don't usually use the robot (i.e. myself) have no idea about what should be tested and would require a lot of wasted time every time, waiting for a slot of time free on the robot, getting everything set up and I at this point they wouldn't know which tests to run, except for the unit test that are already tested.
Moreover there are rare, but still very welcome patches coming from outside IIT, and we can't force the authors of these patches to do these tests.

The fact that our robots run on the `devel` branch and are used at the same time by developers, by researchers and for demos, and that everyone is allowed to upgrade the whole stack and to leave it in whatever state they want is in my opinion just wrong.
What we could IMO enforce is that there should be one person in charge of each robot and nobody else is allowed to upgrade it...

**Comment by drdanz** [comment]
Meanwhile all checks have passed, I'm merging this...

**Comment by drdanz** [comment]
@traversaro The `FindEigen3.cmake` file is already available in `cmake/ycm-0.2/3rdparty/` but it is slightly different, which one is the right one? should I update it in YCM as well?

**Comment by traversaro** [comment]
> I don't think that Eigen is tested by Appveyor, therefore I don't know what will happen on MSVC.

https://github.com/robotology/yarp/pull/896/commits/37fdf22270998f08bd3622f56c607109f4173c43 : )

> @traversaro The FindEigen3.cmake file is already available in cmake/ycm-0.2/3rdparty/ but it is slightly different, which one is the right one? should I update it in YCM as well?

I always use the latest from Eigen default branch, see https://bitbucket.org/eigen/eigen/src/00094bd861b8b6d355890702df7df19d9a3d99de/cmake/FindEigen3.cmake?at=default&fileviewer=file-view-default (so I guess we should update also the one in YCM).

**Comment by drdanz** [comment]
> 37fdf22 : )

Oops, I missed this :+1:

**Comment by pattacini** [comment]
Well, don't want to seem too picky but there're quite important issues here about the way we develop and test software. Let's discuss them this Thursday then.

**Pull request merged**